zstd/zstd.go

```diff
@@ -38,161 +35,91 @@ func (c *CompressionCodec) Name() string { return "zstd" }
-// =============================================================================
+var decPool sync.Pool
 
 type reader struct {
-	reader io.Reader
-	buffer *buffer
-	offset int
+	dec *zstdlib.Decoder
+	err error
 }
 
-func (r *reader) Read(b []byte) (int, error) {
-	if err := r.decompress(); err != nil {
-		return 0, err
-	}
-
-	if r.offset >= len(r.buffer.output) {
-		return 0, io.EOF
+// Close implements the io.Closer interface.
+func (r *reader) Close() error {
+	if r.err != io.ErrClosedPipe {
+		r.err = io.ErrClosedPipe
+		decPool.Put(r.dec)
 	}
-
-	n := copy(b, r.buffer.output[r.offset:])
-	r.offset += n
-	return n, nil
+	return nil
 }
 
-func (r *reader) WriteTo(w io.Writer) (int64, error) {
-	if err := r.decompress(); err != nil {
-		return 0, err
-	}
-
-	if r.offset >= len(r.buffer.output) {
-		return 0, nil
+// Read implements the io.Reader interface.
+func (r *reader) Read(p []byte) (n int, err error) {
+	println("zstd read", len(p))
+	if r.err != nil {
+		return 0, r.err
 	}
-
-	n, err := w.Write(r.buffer.output[r.offset:])
-	r.offset += n
-	return int64(n), err
+	return r.dec.Read(p)
 }
 
-func (r *reader) Close() (err error) {
-	if b := r.buffer; b != nil {
-		r.buffer = nil
-		b.reset()
-		bufferPool.Put(b)
+// WriteTo implements the io.WriterTo interface.
+func (r *reader) WriteTo(w io.Writer) (n int64, err error) {
+	if r.err != nil {
+		return 0, r.err
 	}
-	return
+	return r.dec.WriteTo(w)
 }
 
-func (r *reader) decompress() (err error) {
-	if r.reader == nil {
-		return
-	}
-
-	b := r.buffer
-
-	if _, err = b.readFrom(r.reader); err != nil {
-		return
+// NewWriter implements the kafka.CompressionCodec interface.
+func (c *CompressionCodec) NewWriter(w io.Writer) io.WriteCloser {
+	p := new(writer)
+	if cached := encPool.Get(); cached == nil {
+		p.enc, p.err = zstdlib.NewWriter(w,
+			zstdlib.WithEncoderLevel(c.level))
+	} else {
+		p.enc = cached.(*zstdlib.Encoder)
+		p.enc.Reset(w)
 	}
-
-	r.reader = nil
-	b.output, err = zstd.Decompress(b.output[:cap(b.output)], b.input)
-	return
-}
-
-type writer struct {
-	writer io.Writer
-	buffer *buffer
-	level  int
+	return p
 }
 
-func (w *writer) Write(b []byte) (int, error) {
-	return w.buffer.write(b)
-}
+var encPool sync.Pool
 
-func (w *writer) ReadFrom(r io.Reader) (int64, error) {
-	return w.buffer.readFrom(r)
+type writer struct {
+	enc *zstdlib.Encoder
+	err error
 }
 
-func (w *writer) Close() (err error) {
-	if b := w.buffer; b != nil {
-		w.buffer = nil
-
-		b.output, err = zstd.CompressLevel(b.output[:cap(b.output)], b.input, w.level)
-		if err == nil {
-			_, err = w.writer.Write(b.output)
-		}
-
-		b.reset()
-		bufferPool.Put(b)
+// Close implements the io.Closer interface.
+func (w *writer) Close() error {
+	if w.err == io.ErrClosedPipe {
+		return nil
 	}
-	return
-}
-
-type buffer struct {
-	input  []byte
-	output []byte
-}
-
-func (b *buffer) reset() {
-	b.input = b.input[:0]
-	b.output = b.output[:0]
+	w.err = io.ErrClosedPipe
+	encPool.Put(w.enc)
```

Comment:
Defer this, otherwise someone might pick it up before it is closed.

Reply:
Ping

Reply:
Meh. I don't think a defer should be used with two lines of logic in place.  A swap makes the "unset" a bit less clear. Feel free to correct me @achille-roussel.

Reply:
I think @klauspost was referring to the call to `Close` after the encoder was given back to the pool. If the encoder gets reused before the `Close` call is executed, it would result in race conditions and unpredictable behaviors.

I'd have the same comment as for the reader path here, let's not use the `w.err` field as the sentinel for whether the encoder was closed, because other code paths may be setting this field as well. How about replacing it with something like this?
```go
if w.enc != nil {
    w.enc.Close()
    encPool.Put(w.enc)
    w.enc = nil
    w.err = io.ErrClosedPipe
}
```

Reply:
That's much better indeed.